File: 200-number-of-islands/number-of-islands.java
class Solution {
    int[] delrow={-1,1,0,0};
    int[] delcol={0,0,-1,1};
    public int numIslands(char[][] grid) {
        int n=grid.length;
        int m=grid[0].length;

        int[][] visited= new int[n][m];
        int count=0;
        for(int i=0; i<n ;i++){
            for(int j=0; j<m ; j++){
                if(visited[i][j]==0 && grid[i][j]=='1'){
                    count++;
                    dfs(i, j, visited, grid);
                }
            }
        }
        return count;
    }

    private void dfs(int row, int col, int[][] visited, char[][] grid){
        visited[row][col]=1;
        for(int i=0; i<4; i++ ){
            int nrow= row+delrow[i];
            int ncol= col+delcol[i];
            if(nrow>=0 && nrow<grid.length && ncol>=0 && ncol<grid[0].length && visited[nrow][ncol]==0 && grid[nrow][ncol]=='1'){
                dfs(nrow, ncol, visited, grid);
            }
        }
    }
}

File: 200-number-of-islands/number-of-islands.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        n= len(grid)
        m= len(grid[0])

        visited= [[0]*m for _ in range(n)]

        def bfs(visited, grid, row, col):
            n= len(grid)
            m= len(grid[0])
            visited[row][col]=1
            q= deque()
            q.append((row, col))
            directions= [(1,0),(0,1),(-1,0),(0,-1)]

            while q:
                nr, nc = q.popleft()
                visited[row][col]=1

                for r,c in directions:
                    nrow= r+nr
                    ncol= c+nc

                    if(0<=nrow< n and 0<= ncol <m and grid[nrow][ncol]=="1" and visited[nrow][ncol]==0):
                        visited[nrow][ncol]=1
                        q.append((nrow,ncol))


        res=0
        for i in range(n):
            for j in range(m):
                if visited[i][j]==0 and grid[i][j]=="1":
                    bfs(visited, grid, i, j )
                    res+=1
        return res
        